src/main/java/net/neoforged/neoforge/transfer/storage/Storage.java

```diff
@@ -0,0 +1,140 @@
+     * @return The amount of the resource that was (or would have been, if simulated) inserted.
+     */
+    long insert(int index, T resource, long maxAmount, TransactionContext transaction);
+
+    /**
+     * Inserts a given amount of the resource into the handler. Distribution of the resource is up to the handler.
+     *
+     * @param resource The resource to insert.
+     * @param maxAmount The amount of the resource to insert.
+     * @return The amount of the resource that was (or would have been, if simulated) inserted.
+     */
+    default long insert(T resource, long maxAmount, TransactionContext transaction) {
+        long amount = 0;
+        int slots = size();
+
+        for (int i = 0; i < slots; ++i) {
+            amount += insert(i, resource, maxAmount - amount, transaction);
+            if (amount == maxAmount) break;
+        }
+
+        return amount;
+    }
+
+    /**
+     * Return false if calling {@link #insert} will absolutely always return 0, or true otherwise or in doubt.
+     *
+     * <p>Note: This function is meant to be used by pipes or other devices that can transfer resources to know if
+     * they should interact with this storage at all.
+     */
+    default boolean supportsInsertion() {
+        return true;
+    }
+
+    /**
+     * Extracts a given amount of the resource from the handler at the given index.
+     *
+     * @param index The index to extract the resource from.
+     * @param resource The resource to extract.
+     * @param maxAmount The amount of the resource to extract.
+     * @return The amount of the resource that was (or would have been, if simulated) extracted.
+     */
+    long extract(int index, T resource, long maxAmount, TransactionContext transaction);
+
+    /**
+     * Extracts a given amount of the resource from the handler. Distribution of the resource is up to the handler.
+     *
+     * @param resource The resource to extract.
+     * @param maxAmount The amount of the resource to extract.
+     * @return The amount of the resource that was (or would have been, if simulated) extracted.
+     */
+    default long extract(T resource, long maxAmount, TransactionContext transaction) {
+        long amount = 0;
+        int slots = size();
+
+        for (int i = 0; i < slots; ++i) {
+            amount += extract(i, resource, maxAmount - amount, transaction);
+            if (amount == maxAmount) break;
+        }
+
+        return amount;
+    }
+
+    /**
+     * Return false if calling {@link #extract} will absolutely always return 0, or true otherwise or in doubt.
+     *
+     * <p>Note: This function is meant to be used by pipes or other devices that can transfer resources to know if
+     * they should interact with this storage at all.
+     */
+    default boolean supportsExtraction() {
+        return true;
+    }
+
+    // TODO: is this a good idea? we might need an iterator and an iterable method
```

Comment:
I'm not super convinced by this. What we could maybe do is
```java
default int nextSlot(int slot) {
    return slot+1;
}
```

which is basically iterator-at-home, but it avoids all the allocations that an iterator would otherwise have.

cc @shartte since you brought up the non-empty iterator again.

Reply:
avoiding allocations can be done by just `for (i)` looping the indexes.

imo either return an array, or make it be a consumer like `List.forEach`